File: csharp/Src/Shapes/Base/Models/Rectangle.cs
using Shapes.Base.Commons;

namespace Shapes.Base.Models;

public sealed class Rectangle : IDimensional {
    private readonly double _width;
    private readonly double _height;

    public Rectangle(double width, double height) {
        double dim1 = width > 0 ? width : 1;
        double dim2 = height > 0 ? height : 1;
        _width = dim1;
        _height = dim2;
    }

    public double Area() {
        return _width * _height;
    }

    public double Perimeter() {
        return 2 * (_width + _height);
    }
}

File: haskell/src/Shapes/Base/Models/Rectangle.hs
module Shapes.Base.Models.Rectangle (
    Rectangle (..),
    newRectangle,
) where

import Shapes.Base.Commons

data Rectangle where
    Rectangle :: {width :: !Double, height :: !Double} -> Rectangle
    deriving (Eq, Show)

instance Dimensional Rectangle where
    area (Rectangle width height) = width * height
    perimeter (Rectangle width height) = (width + height) * 2

newRectangle :: Double -> Double -> Rectangle
newRectangle width height = Rectangle (max width 1) (max height 1)
